### PR TITLE
Rewrite tests for RTCConfiguration iceServers

### PR DESCRIPTION
**Pull request by soareschen**
The latest editor's draft have removed the `token` `credentialType` since 20170508. Instead a new `oath` `credentialType` is added. As a result, some of the existing tests for RTCPeerConnection-constructor is no longer valid.

I have extracted the existing tests for iceServers from RTCPeerConnection-constructor.html and rewrite them in RTCConfiguration-iceServers.html. The rewrite includes coverage annotation with current spec, and test helper function to test the same configuration with `new RTCPeerConnection()` and `setConfiguration()`.

The initial commit for this PR points to the tip of tree draft. The main difference is that `NotSupportedError` is thrown instead of `SyntaxError` when unknown URL scheme is provided. (w3c/webrtc-pc#1433). To make ease of the transition, I added a commit 0dc620f that revert the test back to the latest editor draft. Hopefully with this approach, this PR can be merged first without issue, and when new editor's draft is published, commit 0dc620f can be reverted so that the tests can be updated without too much effort.

### CODE REVIEW

**Comment by sideshowbarker** [comment]
w3c-test:mirror

File: webrtc/RTCConfiguration-iceServers.html [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+                 DOMString                          username;
+                 (DOMString or RTCOAuthCredential)  credential;
+                 RTCIceCredentialType               credentialType = "password";
+      };
+   */
+
+  test(() => {
+    const pc = new RTCPeerConnection();
+    assert_equals(pc.getConfiguration().iceServers, undefined);
+  }, 'new RTCPeerConnection() should have default configuration.iceServers of []');
```
Comment (rwaldron) [inline webrtc/RTCConfiguration-iceServers.html]: This: 

> 'new RTCPeerConnection() should have default configuration.iceServers of []'

Does not reflect the test: 

> `assert_equals(pc.getConfiguration().iceServers, undefined);`

If `iceServers` is an empty array, it would not equal `undefined`

Comment (rwaldron) [inline webrtc/RTCConfiguration-iceServers.html]: Also consider: https://heycam.github.io/webidl/#es-sequence

File: webrtc/RTCConfiguration-iceServers.html [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+
+  config_test(makePc => {
+    assert_throws(new TypeError(), () =>
+      makePc({ iceServers: null }));
+  }, '{ iceServers: null } should throw TypeError');
+
+  config_test(makePc => {
+    const pc = makePc({ iceServers: undefined });
+    assert_equals(pc.getConfiguration().iceServers, undefined);
+  }, '{ iceServers: undefined } should succeed');
```
Comment (rwaldron) [inline webrtc/RTCConfiguration-iceServers.html]: See comment above

File: webrtc/RTCConfiguration-iceServers.html [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+  // token credentialType was removed from the spec since 20170508
+  config_test(makePc => {
+    assert_throws(new TypeError(), () =>
+      makePc({ iceServers: [{
+        urls: [],
+        credentialType: 'token'
+      }] }));
+  }, 'with credentialType token should throw TypeError');
+
+  // Blink and Gecko fall back to url, but it's not in the spec.
```
Comment (rwaldron) [inline webrtc/RTCConfiguration-iceServers.html]: Then it's wrong and shouldn't be included as a test of spec implementation

File: webrtc/RTCConfiguration-iceServers.html [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+        credential: 'cred'
+      }] }));
+  }, 'with turns server and only credential should throw InvalidAccessError');
+
+  /*
+    4.3.2.  To set a configuration
+      11.3. For each url in server.urls parse url and obtain scheme name.
+        - If the scheme name is not implemented by the browser, throw a SyntaxError.
+        - or if parsing based on the syntax defined in [ RFC7064] and [RFC7065] fails,
+          throw a SyntaxError.
```
Comment (rwaldron) [inline webrtc/RTCConfiguration-iceServers.html]: Unrelated to these tests: this is another misuse of SyntaxError by this spec :(

File: webrtc/RTCConfiguration-iceServers.html [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+      }] }));
+  }, 'with turns server and only username should throw InvalidAccessError');
+
+  config_test(makePc => {
+    assert_throws('InvalidAccessError', () =>
+      makePc({ iceServers: [{
+        urls: 'turns:turn.example.net',
+        credential: 'cred'
+      }] }));
+  }, 'with turns server and only credential should throw InvalidAccessError');
```
Comment (rwaldron) [inline webrtc/RTCConfiguration-iceServers.html]: Thorough!

File: webrtc/RTCConfiguration-iceServers.html [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+    assert_equals(iceServers.length, 1);
+
+    const server = iceServers[0];
+    assert_array_equals(server.urls, ['stun:stun1.example.net']);
+    assert_equals(server.credentialType, 'password');
+    assert_equals(server.username, 'user');
+
+    const { credential } = server;
+    assert_equals(credential.macKey, '');
+    assert_equals(credential.accessToken, '');
```
Comment (rwaldron) [inline webrtc/RTCConfiguration-iceServers.html]: Not one single browser has implemented this correctly :(

File: webrtc/RTCConfiguration-iceServers.html [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+    Untestable
+      4.3.2.  To set a configuration
+        11.7.  Append server to validatedServers.
+
+    Coverage Report
+      Tested         9
+      Not Tested     0
+      Untestable     1
+      Total         10
+   */
```
Comment (rwaldron) [inline webrtc/RTCConfiguration-iceServers.html]: Confirmed. I've updated several sections of the coverage spec.

**Review by rwaldron — CHANGES_REQUESTED** [review]
This appears to have conflicts, please rebase

File: webrtc/RTCConfiguration-iceServers.html (hunk 1 above)
Comment (soareschen) [inline webrtc/RTCConfiguration-iceServers.html]: Oh yeah, I think I was checking whether an optional sequence field in dictionary has default value of `[]`. It is hard to search in WebIDL but I believe the default is still `undefined`. So the description should be updated to reflect the test.

File: webrtc/RTCConfiguration-iceServers.html (hunk 4 above)
Comment (soareschen) [inline webrtc/RTCConfiguration-iceServers.html]: This has been updated in webrtc-pc master branch to throw `NotSupportedError` instead. Unfortunately the editor's draft have not been updated so I put this in a separate commit that is revertable once the next editor's draft is published.

File: webrtc/RTCConfiguration-iceServers.html (hunk 3 above)
Comment (soareschen) [inline webrtc/RTCConfiguration-iceServers.html]: This is one of the original tests written by @foolip. I guess it might be legacy features in the spec that was later removed. Since the test cases still fail in current browsers I decided to keep them during refactoring.

File: webrtc/RTCConfiguration-iceServers.html (hunk 4 above)
Comment (rwaldron) [inline webrtc/RTCConfiguration-iceServers.html]: Oh, that's great news.

**Review by rwaldron — APPROVED** [review]

File: webrtc/RTCConfiguration-iceServers.html (hunk 3 above)
Comment (rwaldron) [inline webrtc/RTCConfiguration-iceServers.html]: If there is no spec for the behavior, then there shouldn't be any tests. Legacy or not. I will file an issue to remove it.

Comment (foolip) [inline webrtc/RTCConfiguration-iceServers.html]: I added this comment in 089c33e7267854a391cfd8ce38929c52425f1252 to explain the `'{ iceServers: [{ url: "stun:stun1.example.net" }] }': new TypeError,` line that followed but which isn't in this file.

It's an historical test, to make sure that some test is failing as long as a formerly-standard thing hasn't been removed from implementations. These are often in historical.html tests. I'm adding one for WebRTC in https://github.com/w3c/web-platform-tests/pull/6686, once that's landed the test for non-support of `url` could be moved there.
